runners/spark/src/main/java/org/apache/beam/runners/spark/util/SideInputStorage.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.spark.util;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.PCollectionView;
+
+/**
+ * Cache deserialized side inputs for executor so every task doesn't need to deserialize them again.
+ * Side inputs are stored in {@link Cache} with 5 minutes expireAfterAccess.
+ */
+class SideInputStorage {
+
+  /** JVM deserialized side input cache. */
+  private static final Cache<Key<?>, Optional<?>> materializedSideInputs =
```

Comment:
I am a bit worried on the possible consequence of a collision of the `Key<view, window>` tuple in particular if a bad implementation of equals is around. This is not relative to this PR but since the state is now static this makes the likelihood of this happening bigger.

Reply:
how do other runners cache side inputs? by which key? this sounds like something the SDK could provide guidance on (@kennknowles)

Reply:
The window should use `windowCoder.structuralValue(window)` which is required to behave identically to a full serialization for shuffle purposes. The view itself is just a tag so it should have good enough equals as-is. There is already caching in the now-donated Dataflow Java worker, if you look through uses and subclasses of `SideInputReader`.

Reply:
I also think you have bigger troubles if you have collision in `view` or `window`(on more places `sparkRunner` relies on that) so I will leave it as it is. Is that ok?